XNAControls/XNAScrollPanel.cs

```diff
@@ -0,0 +1,658 @@
+            bool left = Keyboard.IsKeyHeldDown(Keys.Left);
+            bool right = Keyboard.IsKeyHeldDown(Keys.Right);
+            
+            bool scrollingHorizontally = left != right;
+            bool scrollingVertically = up != down;
+            bool anyValidInput = scrollingHorizontally || scrollingVertically;
+            
+            if (scrollingVertically)
+                HandleScrollKeyDown(gameTime, up ? ScrollUp : ScrollDown);
+            
+            if (scrollingHorizontally)
+                HandleScrollKeyDown(gameTime, left ? ScrollLeft : ScrollRight);
+
+            if (!anyValidInput)
+            {
+                _isScrollingQuickly = false;
+                _timeSinceLastScroll = TimeSpan.Zero;
+                _scrollKeyTime = TimeSpan.Zero;
+            }
+        }
+
+        base.Update(gameTime);
+    }
+
+    protected virtual void HandleScrollKeyDown(GameTime gameTime, Action action)
+    {
+        if (_scrollKeyTime.Equals(TimeSpan.Zero))
+            action();
+
+        WindowManager.SelectedControl = this;
+
+        _scrollKeyTime += gameTime.ElapsedGameTime;
+
+        if (_isScrollingQuickly)
+        {
+            _timeSinceLastScroll += gameTime.ElapsedGameTime;
+
+            if (_timeSinceLastScroll > TimeSpan.FromSeconds(SCROLL_REPEAT_TIME))
+            {
+                _timeSinceLastScroll = TimeSpan.Zero;
+                action();
+            }
+        }
+
+        if (_scrollKeyTime > TimeSpan.FromSeconds(FAST_SCROLL_TRIGGER_TIME) && !_isScrollingQuickly)
+        {
+            _isScrollingQuickly = true;
+            _timeSinceLastScroll = TimeSpan.Zero;
+        }
+    }
+    
+    #endregion
+
+    #region Recalculation methods
+    
+    /// <summary>
+    /// Readjusts the inner panel size using coords of its children.
+    /// </summary>
+    protected virtual void RecalculateContentSize()
+    {
+        // TODO profile and perhaps optimize this via sorted array of max control sizes
+        Point contentSize = ContentPanel.Children
+            .Select(c => new Point(c.Right, c.Bottom))
+            .Aggregate(Point.Zero, (accumulated, next) 
+                => new Point(Math.Max(accumulated.X, next.X), Math.Max(accumulated.Y, next.Y)));
+
+        ContentSize = contentSize + OverscrollMargin;
+    }
+
+    /// <summary>
+    /// Readjusts the scrollbar sizes, max values etc.
+    /// </summary>
+    protected virtual void RecalculateScrollbars()
+    {
+        if (Width == 0 || Height == 0)
+            return;
+        
+        int border = DrawBorders ? 1 : 0;
+        int border2x = border * 2;
+
+        #region Visibility calculations
+
+        Point viewSizeNoScrollbars = ClientRectangle.Size;
+        Point viewSizeWithScrollbars = new()
+        {
+            X = Width
+                - VerticalScrollBar.ScrollWidth 
+                - border,
+            Y = Height
+                - HorizontalScrollBar.ScrollHeight 
+                - border,
```

Comment:
Modern monitors typically have at least 1920 pixels of horizontal space, I suggest putting some of it to use.

```suggestion
            X = Width - VerticalScrollBar.ScrollWidth - border,
            Y = Height - HorizontalScrollBar.ScrollHeight - border,
```